docs: adding dynamic channel page (#1959)
asyncapi/website: pages/docs/concepts/asyncapi-document/dynamic-channel-address.md

---
title: Dynamic Channel Address
weight: 80
---

Dynamic channel addresses are used to specify dynamic parts of a channel name, which becomes particularly useful when you want to use and reuse parameters in the channel name. It provides flexibility in how you structure your event-driven interfaces.

Here is the diagram explaining dynamic channel address:

```mermaid
graph TD
A[AsyncAPI]
B[Dynamic Channel Address]
D[Variable Usage in Channel Address]
E[Adapted to Multiple APIs and Protocols]
F[Reused Across Channels]

style B fill:#47BCEE,stroke:#47BCEE;

A --> B
B --> D
D --> E
D --> F
```

> The flowchart shows how dynamic channel addresses allow flexible event-driven interfaces by using and reusing parameters in channel names, adapting well to various APIs and protocols.

Reviewer: We do not use the word "flowchart" for diagrams in our docs. Let's remember to be consistent, and please update all mentions to say "diagram."

Here is an example of dynamic channel address:

```yml
userSignedUp:
  address: 'user.signedup'
  messages:
    userSignedUp:
      $ref: '#/components/messages/userSignedUp'
```

This document defines a dynamic channel address for a `userSignedUp` event message, making it easy to include and reuse specific details in the channel address.

> ## Parameter Context

Reviewer: We observe sentence capitalization for headers in our docs; please update all headers accordingly. 😄

---
title: Dynamic Channel Address
weight: 80
---

> Dynamic channel addresses are used to specify dynamic parts of a channel name, which becomes particularly useful when you want to use and reuse parameters in the channel name. It provides flexibility in how you structure your event-driven interfaces.

Reviewer 2: Changing `name` to `address` to be correct. Also simplifying suggestion:

```suggestion
Dynamic channel addresses specify dynamic parts of a channel address, which becomes particularly useful when you want to use and reuse parameters in a channel address.
```

Here is the diagram explaining dynamic channel address:

```mermaid
      graph TD
      A[AsyncAPI]
      B[Dynamic Channel Address]
      D[Variable Usage in Channel Address]
      E[Adapted to Multiple APIs and Protocols]
      F[Reused Across Channels]

      style B fill:#47BCEE,stroke:#47BCEE;

      A --> B
      B --> D
      D --> E
      D --> F

> ```

Reviewer 2: As mentioned in https://github.com/asyncapi/website/pull/1959#discussion_r1281588954, I still believe this diagram doesn't clarify anything to the user but rather can lead to confusion. I would suggest we remove it.

The diagram shows how dynamic channel addresses allow flexible event-driven interfaces by using and reusing parameters in channel names, adapting well to various APIs and protocols.

Here is an example of dynamic channel address:

```yml
userSignedUp:
  address: 'user.signedup'
  messages:
    userSignedUp:

>       $ref: '#/components/messages/userSignedUp'

Reviewer 2: This example is not showing any parameter in the address. You are already providing one example below in https://github.com/asyncapi/website/pull/1959/files#diff-798204f3a6dff3c49407fb1af6e8d8d4a10a5b26e328844fe1e03e17173e3c98R66.

Author: So here shall I remove the code example? Or I should provide a new example?

```

This document defines a dynamic channel address for a `userSignedUp` event message, making it easy to include and reuse specific details in the channel address.

## Parameter context

> In a channel address, there's a map of parameters, which needs to include all the same parameters that are in the main channel address. The names you use for the parameters in this map must be exactly the same as the names you used in the channel address.

Reviewer 2: Short and clear 👍 I like it

Here is a diagram explaining parameter context:

```mermaid
graph TD
    A[Channel Address] --> B{Parameter Map}
    B --> C{Parameters}
    C --> D[Parameter 1]
    C --> E[Parameter 2]
    C --> F[Parameter 3]
    D --> G{Parameter Name}
    E --> G
    F --> G
    A --> G

>     style B fill:#47BCEE,stroke:#47BCEE;

Reviewer 2: As mentioned in https://github.com/asyncapi/website/pull/1959#discussion_r1281588954, I still believe these diagrams don't clarify anything to the user but rather can lead to confusion. I would suggest we remove it.

```

This diagram shows how the channel address includes the same parameters as the channel address, and each parameter's name matches the one used in the address.

> Here is an example of parameter context:

Reviewer 2: What does `context` mean here? I understand the title above but here doesn't have a meaning IMHO. Suggesting a change:
```suggestion
Here is an example of a dynamic channel address and its parameter:
```

```yml
user/{userId}/signup:
  parameters:
    userId:
      description: Id of the user.
      location: $message.payload#/user/id
  subscribe:
    message:
      $ref: "#/components/messages/userSignedUp"
```

> In the above document, `user/{userId}/signedup` is the endpoint where `{userId}` is a path parameter.

Reviewer 2: Nitpick: Using ubiquitous language
```suggestion
In the above document, `user/{userId}/signedup` is the address where `{userId}` is a parameter for that address.
```

## Reusing parameters

> Parameters can be reused. This reuse optimizes the usage of parameters within the event-driven interface by letting you use the parameters again.

Reviewer 2: Nitpick: using ubiquitous language
```suggestion
Parameters can be reused. This reuse optimizes the usage of parameters within the API by letting you use the parameters again.
```

> If there is another message, for example, a `UserUpdated` message, which also requires the `userId`, the parameter can be reused like the following following diagram:

Reviewer 2: ```suggestion
If there is another message, for example, a `UserUpdated` message, which also requires the `userId` parameter, it can be reused like in the following example:
```

```mermaid
graph TD
    subgraph Channels
        userSignedUp["userSignedUp"]
    end

    subgraph Parameters
        userId["userId"]
    end

    userSignedUp -->|requires| userId

    subgraph Reused Parameters
        UserUpdated["UserUpdated"]
    style UserUpdated fill:#47BCEE,stroke:#47BCEE;
    
    end
    UserUpdated -->|reuses| userId
```

> In this diagram, a channel named `userSignedUp` that requires a parameter `userId` .Additionally, the parameter `userId` is reused for another message, `UserUpdated`.

Reviewer 2: As mentioned in https://github.com/asyncapi/website/pull/1959#discussion_r1281588954, I still believe these diagrams don't clarify anything to the user but rather can lead to confusion. I would suggest we remove it.

The example below is more than enough to illustrate

> Here is a code of reusing parameters:

Reviewer 2: ```suggestion
Here is an example of reusing parameters:
```

```yml
channels:
  userSignedUp:
    address: 'user/{userId}/signedup'
    parameters:
      userId:
        description: Id of the user.
    parameters:
      UserUpdated:
        description: Updated Id of the user.    
operations: 
  userSignedUp:
    action: receive
    channel: 
      $ref: '#/channels/userSignedUp'  

> ```

Reviewer 2: This example is not reusing any parameter. Also, I'm confused because you are defining `parameters` value twice.

Here I created an example for you:
```yaml
channels:
  UserSignedUp:
    address: 'user/{userId}/signedup'
    parameters:
      $ref: '#/components/parameters/userId'
  UserUpdated:
    address: 'user/{userId}/updated'
    parameters:
      $ref: '#/components/parameters/userId'
components:
  parameters:
    userId:
      description: Id of the user.
```

> In this AsyncAPI document, the previously defined `userId` parameter is reused to form the channel address for `UserUpdated` message.

Reviewer 2: Just a suggestion:
```suggestion
In the previous example, the `userId` parameter is defined under `components`, and both channels  `UserSignedUp` and `UserUpdated` are reusing it in their address by referencing via `$ref`.
```
